compatibility/bazel_tools/create-daml-app/Main.hs

```diff
@@ -167,83 +181,128 @@ main = withTempDir $ \yarnCache -> do
-            Nothing -> fail ("Could not decode JSON object from " <> path)
-            Just val -> return val
+    BSL.writeFile packageJsonFile (Aeson.encode newPackageJson)
+
+readJsonFile :: FilePath -> IO Aeson.Value
+readJsonFile path = do
+    -- Read file strictly to avoid lock being held when we subsequently write to it.
+    content <- BSL.fromStrict <$> BS.readFile path
+    case Aeson.decode content of
+        Nothing -> error ("Could not decode JSON object from " <> path)
+        Just val -> return val
+
+extractTarGz :: FilePath -> FilePath -> IO ()
+extractTarGz targz outDir = do
+    runConduitRes
+        $ Conduit.sourceFile targz
+        .| Zlib.ungzip
+        .| Tar.untar (restoreFile (\a b -> fail (T.unpack $ a <> b)) outDir)
+
+setupNpmEnv :: FilePath -> [(TsLibrary, FilePath)] -> IO ()
+setupNpmEnv uiDir libs = do
+  forM_ libs $ \(tsLib, path) -> do
+    let name = tsLibraryName tsLib
+    let uiLibPath = uiDir </> name
+    extractTarGz path uiLibPath
+    patchTsDependencies uiDir $ uiLibPath </> "package.json"
+
+-- | Overwrite dependencies to our TypeScript libraries to point to local file dependencies in the
+-- 'ui' directory in the specified package.json file.
+patchTsDependencies :: FilePath -> FilePath -> IO ()
+patchTsDependencies uiDir packageJsonFile = do
+  packageJson0 <- readJsonFile packageJsonFile
+  case packageJson0 of
+    Aeson.Object packageJson ->
+      case HMS.lookup "dependencies" packageJson of
+        Just (Aeson.Object dependencies) -> do
+          let depNames = HMS.keys dependencies
+          -- patch dependencies to point to local files if they are present in the package.json
+          let patchedDeps =
+                HMS.fromList
+                  ([ ( "@daml.js/create-daml-app"
+                     , Aeson.String $
+                       T.pack $
+                       "file:" <> "./daml.js/create-daml-app-0.1.0")
+                   | "@daml.js/create-daml-app" `elem` depNames
+                   ] ++
+                   [ (depName, Aeson.String $ T.pack $ "file:" <> libRelPath)
+                   | tsLib <- allTsLibraries
+                   , let libName = tsLibraryName tsLib
+                   , let libPath = uiDir </> libName
+                   , let libRelPath =
+                           makeRelative (takeDirectory packageJsonFile) libPath
+                   , let depName = T.replace "-" "/" $ T.pack $ "@" <> libName
+                   , depName `elem` depNames
+                   ]) `HMS.union`
+                dependencies
+          let newPackageJson =
+                Aeson.Object $
+                HMS.insert "dependencies" (Aeson.Object patchedDeps) packageJson
+          -- Make sure we have write permissions before writing
+          p <- getPermissions packageJsonFile
+          setPermissions packageJsonFile (setOwnerWritable True p)
+          BSL.writeFile packageJsonFile (Aeson.encode newPackageJson)
+        Nothing -> pure () -- Nothing to patch
```

Comment:
Do we ever expect to hit this case?

Reply:
That would be the rare case when there are actually no dependencies. Not sure what kind of `package.json` files we generate.